File: agent/scheduler/store.py
from __future__ import annotations

import json
import os
import time
from contextlib import suppress
from dataclasses import dataclass, field
from pathlib import Path
from threading import RLock
from typing import Any, Literal
from uuid import uuid4

from filelock import FileLock
from loguru import logger

from .models import SCHEMA_VERSION, SchedulerJob, validate_job_id


class SchedulerStoreCorrupt(RuntimeError):
    pass


@dataclass
class SchedulerStoreData:
    version: int = SCHEMA_VERSION
    jobs: list[SchedulerJob] = field(default_factory=list)

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> SchedulerStoreData:
        jobs = []
        for item in raw.get("jobs") or []:
            if isinstance(item, dict):
                jobs.append(SchedulerJob.from_dict(item))
        return cls(version=int(raw.get("version") or SCHEMA_VERSION), jobs=jobs)

    def to_dict(self) -> dict[str, Any]:
        return {
            "version": int(self.version or SCHEMA_VERSION),
            "jobs": [job.to_dict() for job in self.jobs],
        }


class SchedulerStore:
    """Durable store for local scheduled jobs.

    The store is intentionally limited to persistence and merge semantics. Timer
    execution lives in the scheduler service so tests can verify each layer in
    isolation.
    """

    def __init__(self, root: Path):
        self.root = Path(root).resolve()
        self.scheduler_dir = self.root / "memory" / "scheduler"
        self.jobs_file = self.scheduler_dir / "jobs.json"
        self.action_file = self.scheduler_dir / "action.jsonl"
        self.lock_file = self.scheduler_dir / "scheduler.lock"
        self._last_action_errors: list[dict[str, Any]] = []
        self._json_lock = RLock()
        self._file_lock = FileLock(str(self.lock_file))
        self._last_good: SchedulerStoreData | None = None
        self._ensure()

    def _ensure(self) -> None:
        self.scheduler_dir.mkdir(parents=True, exist_ok=True)
        if not self.jobs_file.exists():
            self._atomic_write_json(self.jobs_file, SchedulerStoreData().to_dict(), fsync=False)

    def load(self, *, merge_actions: bool = True, allow_last_good: bool = True) -> SchedulerStoreData:
        with self._json_lock:
            with self._file_lock:
                try:
                    data = self._read_store()
                except SchedulerStoreCorrupt:
                    if allow_last_good and self._last_good is not None:
                        return self._last_good
                    raise
                if merge_actions:
                    data = self._merge_actions(data)
                    self._last_good = data
                return data

    def save(self, data: SchedulerStoreData, *, fsync: bool = False) -> None:
        with self._json_lock:
            with self._file_lock:
                payload = data.to_dict()
                self._atomic_write_json(self.jobs_file, payload, fsync=fsync)
                self._last_good = SchedulerStoreData.from_dict(payload)

    def list_jobs(self, *, include_disabled: bool = True) -> list[SchedulerJob]:
        jobs = self.load().jobs
        if not include_disabled:
            jobs = [job for job in jobs if job.enabled]
        return sorted(jobs, key=lambda job: job.state.next_run_at_ms or float("inf"))

    def diagnostics(self) -> dict[str, Any]:
        corrupt_files = sorted(
            self.scheduler_dir.glob("action.corrupt-*.jsonl"),
            key=lambda path: path.stat().st_mtime,
            reverse=True,
        )
        return {
            "jobsFile": self.jobs_file.as_posix(),
            "actionFile": self.action_file.as_posix(),
            "lastActionErrors": list(self._last_action_errors[-20:]),
            "corruptActionFiles": [
                {
                    "path": path.as_posix(),
                    "bytes": path.stat().st_size,
                    "updatedAt": path.stat().st_mtime,
                }
                for path in corrupt_files[:10]
            ],
        }

    def get_job(self, job_id: str) -> SchedulerJob | None:
        safe = validate_job_id(job_id)
        return next((job for job in self.load().jobs if job.id == safe), None)

    def upsert_job(self, job: SchedulerJob) -> SchedulerJob:
        with self._json_lock:
            data = self.load()
            jobs = []
            replaced = False
            for current in data.jobs:
                if current.id == job.id:
                    jobs.append(job)
                    replaced = True
                else:
                    jobs.append(current)
            if not replaced:
                jobs.append(job)
            data.jobs = jobs
            self.save(data)
        return job

    def remove_job(self, job_id: str) -> SchedulerJob | None:
        safe = validate_job_id(job_id)
        with self._json_lock:
            data = self.load()
            removed = next((job for job in data.jobs if job.id == safe), None)
            if removed is None:
                return None
            data.jobs = [job for job in data.jobs if job.id != safe]
            self.save(data)
            return removed

    def append_action(
        self,
        action: Literal["add", "update", "delete"],
        *,
        job: SchedulerJob | None = None,
        job_id: str | None = None,
    ) -> None:
        if action in {"add", "update"} and job is None:
            raise ValueError(f"job is required for action={action}")
        if action == "delete" and not job_id:
            raise ValueError("job_id is required for action=delete")
        payload: dict[str, Any] = {"action": action}
        if job is not None:
            payload["job"] = job.to_dict()
        if job_id:
            payload["jobId"] = validate_job_id(job_id)
        self.scheduler_dir.mkdir(parents=True, exist_ok=True)
        with self._json_lock:
            with self._file_lock:
                with self.action_file.open("a", encoding="utf-8") as f:
                    f.write(json.dumps(payload, ensure_ascii=False) + "\n")

    def _read_store(self) -> SchedulerStoreData:
        try:
            raw = json.loads(self.jobs_file.read_text(encoding="utf-8") or "{}")
            if not isinstance(raw, dict):
                raise ValueError("scheduler store root must be an object")
            data = SchedulerStoreData.from_dict(raw)
            self._last_good = data
            return data
        except Exception as exc:
            backup = self.jobs_file.with_suffix(
                self.jobs_file.suffix + f".corrupt-{int(time.time())}-{uuid4().hex[:8]}"
            )
            with suppress(OSError):
                if self.jobs_file.exists():
                    self.jobs_file.rename(backup)
            raise SchedulerStoreCorrupt(
                f"scheduler store at {self.jobs_file} is corrupt; preserved at {backup}"
            ) from exc

    def _merge_actions(self, data: SchedulerStoreData) -> SchedulerStoreData:
        if not self.action_file.exists():
            return data
        jobs = {job.id: job for job in data.jobs}
        changed = False
        corrupt_records: list[dict[str, Any]] = []
        with self.action_file.open("r", encoding="utf-8") as f:
            for line_no, line in enumerate(f, start=1):
                raw_line = line.rstrip("\n")
                line = raw_line.strip()
                if not line:
                    continue
                try:
                    action = json.loads(line)
                    if not isinstance(action, dict):
                        raise ValueError("action log row must be an object")
                    kind = action.get("action")
                    if kind in {"add", "update"}:
                        job = SchedulerJob.from_dict(action.get("job") or {})
                        jobs[job.id] = job
                        changed = True
                    elif kind == "delete":
                        job_id = validate_job_id(str(action.get("jobId") or action.get("job_id") or ""))
                        if job_id in jobs:
                            jobs.pop(job_id, None)
                            changed = True
                    else:
                        raise ValueError(f"unknown scheduler action: {kind!r}")
                except Exception as exc:
                    logger.warning("Invalid scheduler action log line {}: {}", line_no, exc)
                    corrupt_records.append({
                        "line": line_no,
                        "error": str(exc),
                        "raw": raw_line,
                    })
        if corrupt_records:
            self._write_corrupt_actions(corrupt_records)
            self._last_action_errors = corrupt_records
        if not changed and not corrupt_records:
            return data
        merged = data
        if changed:
            merged = SchedulerStoreData(version=data.version, jobs=list(jobs.values()))
            self._atomic_write_json(self.jobs_file, merged.to_dict(), fsync=False)
        self.action_file.write_text("", encoding="utf-8")
        return merged

    def _write_corrupt_actions(self, records: list[dict[str, Any]]) -> Path:
        path = self.scheduler_dir / f"action.corrupt-{int(time.time())}-{uuid4().hex[:8]}.jsonl"
        with path.open("w", encoding="utf-8") as f:
            for record in records:
                f.write(json.dumps(record, ensure_ascii=False) + "\n")
        return path

    @staticmethod
    def _atomic_write_json(path: Path, payload: dict[str, Any], *, fsync: bool = False) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        tmp = path.with_name(f".{path.name}.{uuid4().hex}.tmp")
        try:
            with tmp.open("w", encoding="utf-8") as f:
                f.write(json.dumps(payload, ensure_ascii=False, indent=2) + "\n")
                if fsync:
                    f.flush()
                    os.fsync(f.fileno())
            tmp.replace(path)
            if fsync:
                with suppress(PermissionError):
                    fd = os.open(str(path.parent), os.O_RDONLY)
                    try:
                        os.fsync(fd)
                    finally:
                        os.close(fd)
        except BaseException:
            tmp.unlink(missing_ok=True)
            raise

File: packages/core/src/scheduler/store.ts
import { randomUUID } from 'node:crypto'
import {
  appendFileSync,
  copyFileSync,
  existsSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  renameSync,
  statSync,
  writeFileSync,
} from 'node:fs'
import { basename, dirname, join } from 'node:path'
import { SCHEMA_VERSION, SchedulerJob, validateJobId } from './models'

export class SchedulerStoreCorrupt extends Error {}

export class SchedulerStoreData {
  version: number
  jobs: SchedulerJob[]
  constructor(opts: { version?: number; jobs?: SchedulerJob[] } = {}) {
    this.version = opts.version ?? SCHEMA_VERSION
    this.jobs = opts.jobs ?? []
  }
  static fromDict(raw: Record<string, any>): SchedulerStoreData {
    const jobs = (raw.jobs ?? []).filter(isObject).map(SchedulerJob.fromDict)
    return new SchedulerStoreData({
      version: Number(raw.version || SCHEMA_VERSION),
      jobs,
    })
  }
  toDict(): Record<string, unknown> {
    return {
      version: this.version || SCHEMA_VERSION,
      jobs: this.jobs.map((job) => job.toDict()),
    }
  }
}

export class SchedulerStore {
  readonly root: string
  readonly schedulerDir: string
  readonly jobsFile: string
  readonly actionFile: string
  readonly lockFile: string
  private lastActionErrors: Array<Record<string, unknown>> = []
  private lastGood: SchedulerStoreData | null = null

  constructor(root: string) {
    this.root = root
    this.schedulerDir = join(root, 'scheduler')
    this.jobsFile = join(this.schedulerDir, 'jobs.json')
    this.actionFile = join(this.schedulerDir, 'action.jsonl')
    this.lockFile = join(this.schedulerDir, 'scheduler.lock')
    mkdirSync(this.schedulerDir, { recursive: true })
    this.copyLegacyFilesIfNeeded()
    if (!existsSync(this.jobsFile))
      this.atomicWriteJson(this.jobsFile, new SchedulerStoreData().toDict())
  }

  load(
    opts: { mergeActions?: boolean; allowLastGood?: boolean } = {},
  ): SchedulerStoreData {
    const mergeActions = opts.mergeActions ?? true
    const allowLastGood = opts.allowLastGood ?? true
    let data: SchedulerStoreData
    try {
      data = this.readStore()
    } catch (error) {
      if (allowLastGood && this.lastGood) return this.lastGood
      throw error
    }
    if (mergeActions) data = this.mergeActions(data)
    this.lastGood = data
    return data
  }

  save(data: SchedulerStoreData): void {
    this.atomicWriteJson(this.jobsFile, data.toDict())
    this.lastGood = SchedulerStoreData.fromDict(
      data.toDict() as Record<string, any>,
    )
  }

  listJobs(opts: { includeDisabled?: boolean } = {}): SchedulerJob[] {
    let jobs = this.load().jobs
    if (opts.includeDisabled === false) jobs = jobs.filter((job) => job.enabled)
    return jobs
      .slice()
      .sort(
        (a, b) =>
          (a.state.next_run_at_ms ?? Infinity) -
          (b.state.next_run_at_ms ?? Infinity),
      )
  }

  getJob(jobId: string): SchedulerJob | null {
    const safe = validateJobId(jobId)
    return this.load().jobs.find((job) => job.id === safe) ?? null
  }

  upsertJob(job: SchedulerJob): SchedulerJob {
    const data = this.load()
    const jobs = data.jobs.filter((item) => item.id !== job.id)
    jobs.push(job)
    data.jobs = jobs
    this.save(data)
    return job
  }

  removeJob(jobId: string): SchedulerJob | null {
    const safe = validateJobId(jobId)
    const data = this.load()
    const removed = data.jobs.find((job) => job.id === safe) ?? null
    if (!removed) return null
    data.jobs = data.jobs.filter((job) => job.id !== safe)
    this.save(data)
    return removed
  }

  appendAction(
    action: 'add' | 'update' | 'delete',
    opts: { job?: SchedulerJob | null; jobId?: string | null } = {},
  ): void {
    if ((action === 'add' || action === 'update') && !opts.job)
      throw new Error(`job is required for action=${action}`)
    if (action === 'delete' && !opts.jobId)
      throw new Error('job_id is required for action=delete')
    const payload: Record<string, unknown> = { action }
    if (opts.job) payload.job = opts.job.toDict()
    if (opts.jobId) payload.jobId = validateJobId(opts.jobId)
    appendFileSync(this.actionFile, JSON.stringify(payload) + '\n', 'utf8')
  }

  diagnostics(): Record<string, unknown> {
    const corrupt = existsSync(this.schedulerDir)
      ? readdirSync(this.schedulerDir)
          .filter(
            (name) =>
              name.startsWith('action.corrupt-') && name.endsWith('.jsonl'),
          )
          .sort()
          .reverse()
      : []
    return {
      jobsFile: this.jobsFile,
      actionFile: this.actionFile,
      lastActionErrors: this.lastActionErrors.slice(-20),
      corruptActionFiles: corrupt.slice(0, 10).map((name) => {
        const path = join(this.schedulerDir, name)
        const st = statSync(path)
        return { path, bytes: st.size, updatedAt: st.mtimeMs / 1000 }
      }),
    }
  }

  private readStore(): SchedulerStoreData {
    try {
      const raw = JSON.parse(readFileSync(this.jobsFile, 'utf8') || '{}')
      if (!isObject(raw))
        throw new Error('scheduler store root must be an object')
      const data = SchedulerStoreData.fromDict(raw)
      this.lastGood = data
      return data
    } catch (error) {
      const backup = `${this.jobsFile}.corrupt-${Math.trunc(Date.now() / 1000)}-${randomUUID().replace(/-/g, '').slice(0, 8)}`
      if (existsSync(this.jobsFile)) {
        try {
          renameSync(this.jobsFile, backup)
        } catch {
          /* ignore */
        }
      }
      throw new SchedulerStoreCorrupt(
        `scheduler store at ${this.jobsFile} is corrupt; preserved at ${backup}`,
        { cause: error },
      )
    }
  }

  private copyLegacyFilesIfNeeded(): void {
    const legacyDir = join(this.root, 'memory', 'scheduler')
    for (const name of ['jobs.json', 'action.jsonl']) {
      const source = join(legacyDir, name)
      const dest = join(this.schedulerDir, name)
      if (existsSync(dest) || !existsSync(source)) continue
      try {
        copyFileSync(source, dest)
      } catch {
        /* non-destructive best effort */
      }
    }
  }

  private mergeActions(data: SchedulerStoreData): SchedulerStoreData {
    if (!existsSync(this.actionFile)) return data
    const jobs = new Map(data.jobs.map((job) => [job.id, job]))
    let changed = false
    const corruptRecords: Array<Record<string, unknown>> = []
    const lines = readFileSync(this.actionFile, 'utf8').split('\n')
    lines.forEach((rawLine, index) => {
      const line = rawLine.trim()
      if (!line) return
      try {
        const action = JSON.parse(line)
        if (!isObject(action))
          throw new Error('action log row must be an object')
        const kind = action.action
        if (kind === 'add' || kind === 'update') {
          const job = SchedulerJob.fromDict(
            isObject(action.job) ? action.job : {},
          )
          jobs.set(job.id, job)
          changed = true
        } else if (kind === 'delete') {
          const jobId = validateJobId(
            String(action.jobId ?? action.job_id ?? ''),
          )
          if (jobs.delete(jobId)) changed = true
        } else {
          throw new Error(`unknown scheduler action: ${kind}`)
        }
      } catch (error) {
        corruptRecords.push({
          line: index + 1,
          error: String(error instanceof Error ? error.message : error),
          raw: rawLine,
        })
      }
    })
    if (corruptRecords.length) {
      this.writeCorruptActions(corruptRecords)
      this.lastActionErrors = corruptRecords
    }
    if (!changed && !corruptRecords.length) return data
    const merged = changed
      ? new SchedulerStoreData({
          version: data.version,
          jobs: [...jobs.values()],
        })
      : data
    if (changed) this.atomicWriteJson(this.jobsFile, merged.toDict())
    writeFileSync(this.actionFile, '', 'utf8')
    return merged
  }

  private writeCorruptActions(records: Array<Record<string, unknown>>): string {
    const path = join(
      this.schedulerDir,
      `action.corrupt-${Math.trunc(Date.now() / 1000)}-${randomUUID().replace(/-/g, '').slice(0, 8)}.jsonl`,
    )
    writeFileSync(
      path,
      records.map((record) => JSON.stringify(record)).join('\n') + '\n',
      'utf8',
    )
    return path
  }

  private atomicWriteJson(
    path: string,
    payload: Record<string, unknown>,
  ): void {
    mkdirSync(dirname(path), { recursive: true })
    const tmp = join(
      dirname(path),
      `.${basename(path)}.${randomUUID().replace(/-/g, '')}.tmp`,
    )
    writeFileSync(tmp, JSON.stringify(payload, null, 2) + '\n', 'utf8')
    renameSync(tmp, path)
  }
}

function isObject(value: unknown): value is Record<string, any> {
  return Boolean(value && typeof value === 'object' && !Array.isArray(value))
}
